src/internal/operators/catchError.ts

```diff
@@ -111,44 +111,29 @@ export function catchError<T, O extends ObservableInput<any>>(
   return (source: Observable<T>) =>
     lift(source, function (this: Subscriber<T>, source: Observable<T>) {
       const subscriber = this;
-      const subscription = new Subscription();
-      let innerSub: Subscription | null = null;
-      let syncUnsub = false;
-      let handledResult: Observable<ObservedValueOf<O>>;
+      let nextSource: Observable<T> | null = null;
+      let subscription: Subscription;
 
-      const handleError = (err: any) => {
-        try {
-          handledResult = from(selector(err, catchError(selector)(source)));
-        } catch (err) {
-          subscriber.error(err);
-          return;
+      const tryToSwapSubscription = () => {
+        if (subscription && nextSource && !subscriber.closed) {
+          subscription = nextSource.subscribe(subscriber);
         }
       };
 
-      innerSub = source.subscribe(
+      subscription = source.subscribe(
         new CatchErrorSubscriber(subscriber, (err) => {
-          handleError(err);
-          if (handledResult) {
-            if (innerSub) {
-              innerSub.unsubscribe();
-              innerSub = null;
-              subscription.add(handledResult.subscribe(subscriber));
-            } else {
-              syncUnsub = true;
-            }
+          try {
+            nextSource = from(selector(err, catchError(selector)(source)));
+            tryToSwapSubscription();
+          } catch (selectorErr) {
+            subscriber.error(selectorErr);
           }
         })
       );
 
-      if (syncUnsub) {
-        innerSub.unsubscribe();
-        innerSub = null;
-        subscription.add(handledResult!.subscribe(subscriber));
-      } else {
-        subscription.add(innerSub);
-      }
+      tryToSwapSubscription();
 
-      return subscription;
+      return () => subscription.unsubscribe();
```

Comment:
You should just return the `subscription` here, TBH. In master, we only just recently stopped wrapping these sorts of functions in Subscriptions. But either way, you're wrapping a subscription in a function, which isn't necessary.

Reply:
Right! I thought that it was necessary because the subscription is being reassigned on the `tryToSwapSubscription` function. However, now that I have a closer look at it, that reassignment is not necessary. Nice!